web/src/hooks/useRunData.ts

```diff
@@ -0,0 +1,85 @@
+'use client';
+
+import { useState, useEffect } from 'react';
+import { useRunVerifier } from './useStoredRunVerifier';
+import { RunVerifier } from '@/types';
+import * as stravaUtils from '@/utils/strava';
+import { StravaActivity } from '@/utils/strava';
+
+const useRunData = () => {
+  const { verifier, secret, expiry, updateVerifierAndSecret } = useRunVerifier();
+
+  const [loading, setLoading] = useState(true);
+
+  // TODO: change to more generic type
+  const [data, setData] = useState<StravaActivity[]>([]);
+  const [error, setError] = useState<unknown | null>(null);
+
+  const connected = verifier !== RunVerifier.None;
+
+  useEffect(() => {
+    // TODO: add others verifiers
+    if (verifier !== RunVerifier.Strava) {
+      return;
+    }
+    if (!secret) return;
+
+    // access token not expired, do nothing
+    if (expiry > Date.now() / 1000) return;
+
+    const updateAccessToken = async () => {
+      console.log('Updating access token!');
+      // expired, wait for refresh and expiry to update
+      const { refreshToken } = stravaUtils.splitSecret(secret);
+
+      const { accessToken: newAccessToken, expiry: newExpiry } =
+        await stravaUtils.refreshAccessToken(refreshToken);
+
+      const newSecret = stravaUtils.joinSecret(newAccessToken, refreshToken);
+      // update secret
+      updateVerifierAndSecret(RunVerifier.Strava, newSecret, newExpiry);
+    };
+
+    updateAccessToken().catch(console.error);
+  }, [expiry, secret, updateVerifierAndSecret, verifier]);
+
+  // fetch data when access token are updated and not expired!
+  useEffect(() => {
+    // TODO: add others verifiers
+    if (verifier !== RunVerifier.Strava) {
+      return;
+    }
+    if (!secret) return;
+    if (expiry < Date.now() / 1000) {
```

Comment:
Just want to keep a note here. I am thinking about if there is any possible way to give the two useEffect in this hook a better readability, the if statements seems a bit confusing in the beginning.